irc-proto/src/prefix.rs

```diff
@@ -0,0 +1,204 @@
+//! A module providing an enum for a message prefix.
+use std::borrow::ToOwned;
+use std::string;
+use std::str::FromStr;
+use std::fmt;
+
+/// The Prefix indicates "the true origin of the message", according to the server.
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum Prefix {
+    /// servername, e.g. collins.mozilla.org
+    ServerName(String),
+    /// nickname [ ["!" username] "@" hostname ]
+    /// i.e. Nickname(nickname, username, hostname)
+    /// Any of the strings may be ""
+    Nickname(String, String, String),
+}
+
+impl Prefix {
+    /// Creates a prefix by parsing a string.
+    ///
+    /// # Example
+    /// ```
+    /// # extern crate irc;
+    /// # use irc::client::prelude::*;
+    /// # fn main() {
+    /// Prefix::new_from_str("nickname!username@hostname");
+    /// Prefix::new_from_str("example.com");
+    /// # }
+    /// ```
+    pub fn new_from_str(s: &str) -> Prefix {
+        #[derive(Copy, Clone, Eq, PartialEq)]
+        enum Active {
+            Name = 0,
+            User = 1,
+            Host = 2,
+        }
+
+        let mut name = String::new();
+        let mut user = String::new();
+        let mut host = String::new();
+        let mut active = Active::Name;
+
+        for c in s.chars() {
+            match c {
+                // We consider the '.' to be a ServerName except if a ! has already
+                // been encountered.
+                '.' if active == Active::Name => return Prefix::ServerName(s.to_owned()),
+
+                '!' if active == Active::Name => {
+                    active = Active::User;
+                },
+
+                // The '@' is not special until we've started the username
+                // portion
```

Comment:
According to the grammar, the hostname component can be present even if the username component is absent, though not vice versa, and `@` would indicate the start of a hostname regardless of the presence or absence of a username. (Usernames can't contain `@`.)